Piwo.class.php

```diff
@@ -7,29 +10,55 @@ public static function onParserSetup( &$parser ) {
 		//the "python" magic word with execPy()
 		$parser->setFunctionHook( 'piwo', 'Piwo::execPy', Parser::SFH_OBJECT_ARGS );
 	}
+
+  /**
+   * write the given content to the given filePath
+   */
+  public static function writeStringToFile($filePath,$content) {
+		$f = fopen($filePath, "w");
+		fwrite( $f, $content );
+		fclose( $f );
+		unset($f);
+  }
+
 	// Render the output of {{#python:gram}}.
 	public static function execPy( $parser, $frame, $params ) {
 		//The inputs should contain a gram name and sys.argv
 		//The output should also be wikitext.
 		$page = WikiPage::factory( Title::newFromText( 'Gram:' . $frame->expand( $params[0] ) ) );
 		$name = $frame->expand( $params[0] );
+		$base="/tmp/";
+	  $pyCode= $base . $name . ".py";
+    $pyError= $base . $name . ".error";
 		$content = $page->getContent()->getNativeData() . '';
-		$content = "import sys\nsys.path.append('" . getcwd() . "/extensions/Piwo')" . (((strpos($content, "from mw import") or strpos($content, "import mw")) === false) ? "\nimport mw" : "") . "\nsys.stderr = open('/tmp/" . $name . ".error', 'w')\ndel sys\n" . $content . "\nimport sys\nsys.stderr.close()\ndel sys";
-		$f = fopen("/tmp/" . $name . ".py", "w");
-		fwrite( $f, $content );
-		fclose( $f );
-		unset($f);
+    self::writeStringToFile($pyCode,$content);
+		$python="python3";
+		$cmd=[$python,$pyCode];
 		foreach ($params as &$i) {
-			$i = escapeshellarg( $frame->expand( $i ) );
+			 if ($i>0) $cmd[] = escapeshellarg( $frame->expand( $i ) );
 		}
-		unset($i);
-		$output = shell_exec("python3 /tmp/" . $name . '.py ' . implode(' ', array_slice($params, 1)));
-		if ($output === null) { //$output = file_get_contents("/var/tmp/test");
-			$output = '<pre class="error">' . htmlspecialchars( file_get_contents("/tmp/" . $name . ".error") ) . '</pre>';
+		unset ($i);
```

Comment:
No reason to change this.
```suggestion
		unset($i);
```